port/port_posix.h

```diff
@@ -93,7 +93,7 @@ constexpr bool kLittleEndian = PLATFORM_IS_LITTLE_ENDIAN;
 
 class CondVar;
 
-class Mutex {
+class Mutex : public MutexBase {
```

Comment:
Hmm, now mutexes and cond vars are going to get vtables. And there could be creep where more and more accesses end up going through the base class, potentially incurring vcall overheads.

I think I'm more in favor of using Mutex and CondVar as forward-declared in the public API.

Reply:
Considerations of a base class:
- It makes the `Env::TimedWait()` interface implementable publicly
- We do not need to have separate functions for `InstrumentedCondVar` vs. `CondVar`
- I was thinking pessimistic transaction DB's lock table might be sensitive to this overhead, but realized it already has vtable+virtual call overhead (maybe these interfaces can be consolidated) - https://github.com/facebook/rocksdb/blob/ba597514309b686d8addb59616f067d5522186b7/include/rocksdb/utilities/transaction_db_mutex.h#L14-L19 
- Optimizations (assuming LTO) should work well considering there'll likely be one obvious concrete type at a given `Env::TimedWait()` callsite

Reply:
I think it's ok to forward declare now; we can change it later if we need to in order to support `InstrumentedCondVar` or public extensibility. Since there's no way to extend it externally, we should be able to make such an API change.